internal/contour/endpointstranslator.go

```diff
@@ -97,27 +118,33 @@ func (e *EndpointsTranslator) Query(names []string) []proto.Message {
 
 func (*EndpointsTranslator) TypeURL() string { return resource.EndpointType }
 
-func (e *EndpointsTranslator) addEndpoints(ep *v1.Endpoints) {
-	e.recomputeClusterLoadAssignment(nil, ep)
-}
+// Add adds an entry to the cache. If a ClusterLoadAssignment with the same
+// name exists, it is replaced.
+func (e *EndpointsTranslator) Add(a *v2.ClusterLoadAssignment) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
 
-func (e *EndpointsTranslator) updateEndpoints(oldep, newep *v1.Endpoints) {
-	if len(newep.Subsets) == 0 && len(oldep.Subsets) == 0 {
-		// if there are no endpoints in this object, and the old
-		// object also had zero endpoints, ignore this update
-		// to avoid sending a noop notification to watchers.
-		return
+	if e.entries == nil {
+		e.entries = make(map[string]*v2.ClusterLoadAssignment)
 	}
-	e.recomputeClusterLoadAssignment(oldep, newep)
+	e.entries[a.ClusterName] = a
+	e.Notify(a.ClusterName)
 }
 
-func (e *EndpointsTranslator) removeEndpoints(ep *v1.Endpoints) {
-	e.recomputeClusterLoadAssignment(ep, nil)
+// Remove removes the named entry from the cache. If the entry
+// is not present in the cache, the operation is a no-op.
+func (e *EndpointsTranslator) Remove(name string) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	delete(e.entries, name)
+	e.Notify(name)
 }
 
 // recomputeClusterLoadAssignment recomputes the EDS cache taking into account old and new endpoints.
-func (e *EndpointsTranslator) recomputeClusterLoadAssignment(oldep, newep *v1.Endpoints) {
-	// skip computation if either old and new services or endpoints are equal (thus also handling nil)
+func recomputeClusterLoadAssignment(e *EndpointsTranslator, oldep, newep *v1.Endpoints) {
```

Comment:
Any reason to not make this part of `EndpointsTranslator` when you wouldn't need to pass in the ref each time to this method?

Reply:
This is to emphasize that only public methods of `EndpointsTranslator` are used. This makes it clearer (IMHO) that no locking is required here. When this is a member function, the reader has to think about whether it directly accesses internal fields (which would be legal but not conventional) or should hold locks.